File: forthic-rs/src/tokenizer.rs
use itertools::Itertools;
use lazy_static::lazy_static;
use std::collections::HashSet;

use crate::errors::InterpreterError;
use crate::interpreter_error;
use crate::token::Token;

lazy_static! {
    static ref QUOTE_CHARS: HashSet<char> = HashSet::from(['"', '\'', '^', '\x16']);
    static ref WHITESPACE_CHARS: HashSet<char> = HashSet::from([' ', '\t', '\n', '\r', '(', ')']);
}

pub struct Tokenizer {
    source: Vec<char>,
    position: usize,
    token_buffer: Vec<char>,
}

impl Tokenizer {
    /// A Tokenizer is constructed with an input string and returns the next available token on request.
    pub fn new(source: &str) -> Tokenizer {
        Tokenizer {
            source: source.chars().collect_vec(),
            position: 0,
            token_buffer: vec![],
        }
    }

    /// Returns the next token in the input string.
    /// If the end of the string is reached, returns Token::EOS.
    /// If an error occurs, returns an InterpreterError.
    ///
    /// # Examples
    ///
    /// ```
    /// use forthic_rs::tokenizer::Tokenizer;
    /// use forthic_rs::token::Token;
    ///
    /// let mut tokenizer = Tokenizer::new("1 2 3");
    /// assert_eq!(tokenizer.next_token().unwrap(), Token::Word("1".to_string()));
    /// assert_eq!(tokenizer.next_token().unwrap(), Token::Word("2".to_string()));
    /// assert_eq!(tokenizer.next_token().unwrap(), Token::Word("3".to_string()));
    /// assert_eq!(tokenizer.next_token().unwrap(), Token::EOS);
    /// ```
    pub fn next_token(&mut self) -> Result<Token, InterpreterError> {
        self.clear_token_buffer();
        self.transition_from_start()
    }

    fn clear_token_buffer(&mut self) {
        self.token_buffer.clear();
    }

    fn flatten_token_buffer(&mut self) -> String {
        self.token_buffer.iter().collect()
    }

    fn is_whitespace(&self, c: &char) -> bool {
        WHITESPACE_CHARS.contains(c)
    }

    fn is_quote(&self, c: &char) -> bool {
        QUOTE_CHARS.contains(c)
    }

    fn is_triple_quote(&mut self, index: usize, c: &char) -> bool {
        if !self.is_quote(c) {
            return false;
        }

        if index + 2 >= self.source.len() {
            return false;
        }

        self.source[index + 1] == *c && self.source[index + 2] == *c
    }

    fn is_start_memo(&mut self, index: usize) -> bool {
        if index + 1 >= self.source.len() {
            return false;
        }

        self.source[index] == '@' && self.source[index + 1] == ':'
    }

    /// Tokenization is implemented as a state machine. This is the entry point.
    fn transition_from_start(&mut self) -> Result<Token, InterpreterError> {
        while self.position < self.source.len() {
            let c = self.source[self.position];
            self.position += 1;
            if self.is_whitespace(&c) {
                continue;
            } else if c == '#' {
                return self.transition_from_comment();
            } else if c == ':' {
                return self.transition_from_start_definition();
            } else if self.is_start_memo(self.position - 1) {
                // Skip over ":" in "@:"
                self.position += 1;
                return self.transition_from_start_memo();
            } else if c == ';' {
                return Ok(Token::EndDefinition);
            } else if c == '[' {
                return Ok(Token::StartArray);
            } else if c == ']' {
                return Ok(Token::EndArray);
            } else if c == '{' {
                return self.transition_from_gather_module();
            } else if c == '}' {
                return Ok(Token::EndModule);
            } else if self.is_triple_quote(self.position - 1, &c) {
                // Skip over 2nd and 3rd quote chars
                self.position += 2;
                return self.transition_from_gather_triple_quote(c);
            } else if self.is_quote(&c) {
                return self.transition_from_gather_string(c);
            } else {
                // Back up to beginning of word
                self.position -= 1;
                return self.transition_from_gather_word();
            }
        }
        Ok(Token::EOS)
    }

    fn transition_from_comment(&mut self) -> Result<Token, InterpreterError> {
        while self.position < self.source.len() {
            let c = self.source[self.position];
            self.token_buffer.push(c);
            self.position += 1;
            if c == '\n' {
                break;
            }
        }

        Ok(Token::Comment(self.flatten_token_buffer()))
    }

    fn transition_from_start_definition(&mut self) -> Result<Token, InterpreterError> {
        while self.position < self.source.len() {
            let c = self.source[self.position];
            self.position += 1;

            if self.is_whitespace(&c) {
                continue;
            } else {
                self.position -= 1;
                return self.transition_from_gather_definition_name();
            }
        }

        Err(interpreter_error!("Got EOS in START_DEFINITION"))
    }

    fn transition_from_start_memo(&mut self) -> Result<Token, InterpreterError> {
        while self.position < self.source.len() {
            let c = self.source[self.position];
            self.position += 1;

            if self.is_whitespace(&c) {
                continue;
            } else {
                self.position -= 1;
                return self.transition_from_gather_memo_name();
            }
        }

        Err(interpreter_error!("Got EOS in START_MEMO"))
    }

    fn gather_definition_name(&mut self) -> Result<(), InterpreterError> {
        let invalid_def_chars = HashSet::from(['[', ']', '{', '}']);

        while self.position < self.source.len() {
            let c = self.source[self.position];
            self.position += 1;

            if self.is_whitespace(&c) {
                break;
            } else if self.is_quote(&c) {
                return Err(interpreter_error!("Definitions can't have quotes in them"));
            } else if invalid_def_chars.contains(&c) {
                return Err(interpreter_error!(format!(
                    "Definitions can't have {:?} in them",
                    invalid_def_chars
                )));
            } else {
                self.token_buffer.push(c);
            }
        }

        Ok(())
    }

    fn transition_from_gather_definition_name(&mut self) -> Result<Token, InterpreterError> {
        self.gather_definition_name()?;
        Ok(Token::StartDefinition(self.flatten_token_buffer()))
    }

    fn transition_from_gather_memo_name(&mut self) -> Result<Token, InterpreterError> {
        self.gather_definition_name()?;
        Ok(Token::StartMemo(self.flatten_token_buffer()))
    }

    fn transition_from_gather_module(&mut self) -> Result<Token, InterpreterError> {
        while self.position < self.source.len() {
            let c = self.source[self.position];
            self.position += 1;

            if self.is_whitespace(&c) {
                break;
            } else if c == '}' {
                self.position -= 1;
                break;
            } else {
                self.token_buffer.push(c);
            }
        }

        Ok(Token::StartModule(self.flatten_token_buffer()))
    }

    fn transition_from_gather_triple_quote(
        &mut self,
        delimiter: char,
    ) -> Result<Token, InterpreterError> {
        while self.position < self.source.len() {
            let c = self.source[self.position];

            if c == delimiter && self.is_triple_quote(self.position, &c) {
                self.position += 3;
                return Ok(Token::String(self.flatten_token_buffer()));
            } else {
                self.position += 1;
                self.token_buffer.push(c);
            }
        }

        Err(interpreter_error!("Unterminated triple quoted string"))
    }

    fn transition_from_gather_string(
        &mut self,
        delimiter: char,
    ) -> Result<Token, InterpreterError> {
        while self.position < self.source.len() {
            let c = self.source[self.position];
            self.position += 1;

            if c == delimiter {
                return Ok(Token::String(self.flatten_token_buffer()));
            } else {
                self.token_buffer.push(c);
            }
        }

        Err(interpreter_error!(format!(
            "Unterminated string ({}), {}",
            delimiter,
            self.flatten_token_buffer()
        )))
    }

    fn transition_from_gather_word(&mut self) -> Result<Token, InterpreterError> {
        let word_delimiters = HashSet::from([';', '[', ']', '}']);
        while self.position < self.source.len() {
            let c = self.source[self.position];
            self.position += 1;

            if self.is_whitespace(&c) {
                break;
            } else if word_delimiters.contains(&c) {
                self.position -= 1;
                break;
            } else {
                self.token_buffer.push(c);
            }
        }

        Ok(Token::Word(self.flatten_token_buffer()))
    }
}

/// Iterator implementation for Tokenizer
impl Iterator for Tokenizer {
    type Item = Result<Token, InterpreterError>;

    fn next(&mut self) -> Option<Self::Item> {
        match self.next_token() {
            Ok(Token::EOS) => None,
            result => Some(result),
        }
    }
}

File: experimental/forthic-cpp/ForthicLib/Tokenizer.cpp
#include "pch.h"
#include "Tokenizer.h"

Tokenizer::Tokenizer(string& s) : input(s) 
{
	whitespace = " \r\t\n()";
	token_string = "";
	position = 0;
}

Tokenizer::~Tokenizer()
{
}

bool Tokenizer::is_whitespace(char c)
{
	for (unsigned int i = 0; i < whitespace.length(); i++) {
		if (c == whitespace[i]) return true;
	}
	return false;
}

bool Tokenizer::is_quote(char c)
{
	return (c == '"' || c == '\'');
}

bool Tokenizer::IsTripleQuote(int index, char c)
{
	if (!is_quote(c)) return false;
	if (index + 2 >= input.length()) return false;
	return (input[index + 1] == c && input[index + 2] == c);
}

Token Tokenizer::NextToken()
{
	token_string = "";
	return transition_from_START();
}

Token Tokenizer::transition_from_START()
{
	while (position < input.length())
	{
		char c = input[position++];
		if (is_whitespace(c)) continue;
		else if (c == '#') return transition_from_COMMENT();
		else if (c == ':') return transition_from_START_DEFINITION();
		else if (c == ';') return Token(TokenType::END_DEFINITION);
		else if (c == '[') return Token(TokenType::START_ARRAY);
		else if (c == ']') return Token(TokenType::END_ARRAY);
		else if (c == '{') return transition_from_GATHER_MODULE();
		else if (c == '}') return Token(TokenType::END_MODULE);
		else if (IsTripleQuote(position - 1, c))
		{
			position += 2;  // Skip 2nd and 3rd quote chars
			return transition_from_GATHER_TRIPLE_QUOTE_STRING(c);
		}
		else if (is_quote(c))
		{
			return transition_from_GATHER_STRING(c);
		}
		else return transition_from_GATHER_WORD(c);
	}
	return Token(TokenType::EOS);
}

Token Tokenizer::transition_from_COMMENT()
{
	while (position < input.length())
	{
		char c = input[position++];
		if (c == '\n') break;
		token_string += c;
	}
	return Token(TokenType::COMMENT, token_string);
}


Token Tokenizer::transition_from_START_DEFINITION()
{

	while (position < input.length())
	{

		char c = input[position++];
		if (is_whitespace(c)) continue;
		else if (c == '"' || c == '\'') throw "Definition cannot start with a quote";
		else
		{
			position--;
			return transition_from_GATHER_DEFINITION_NAME();
		}
	}

	throw "Got EOS in START_DEFINITION";
}

Token Tokenizer::transition_from_GATHER_DEFINITION_NAME()
{
	while (position < input.length())
	{
		char c = input[position++];
		if (is_whitespace(c)) break;
		else token_string += c;
	}

	return Token(TokenType::START_DEFINITION, token_string);
}

Token Tokenizer::transition_from_GATHER_MODULE()
{
	while (position < input.length())
	{
		char c = input[position++];
		if (is_whitespace(c)) break;
		else if (c == '}')
		{
			position--;
			break;
		}
		else token_string += c;
	}
	return Token(TokenType::START_MODULE, token_string);
}


Token Tokenizer::transition_from_GATHER_TRIPLE_QUOTE_STRING(char delim)
{
	while (position < input.length())
	{
		char c = input[position];
		if (c == delim && IsTripleQuote(position, c))
		{
			position += 3;
			return Token(TokenType::STRING, token_string);
		}
		else
		{
			token_string += c;
			position++;
		}
	}
	throw "Unterminated triple quote string";
}


Token Tokenizer::transition_from_GATHER_STRING(char delim)
{
	while (position < input.length())
	{
		char c = input[position++];
		if (c == delim)  return Token(TokenType::STRING, token_string);
		else             token_string += c;
	}
	throw "Unterminated string";
}


Token Tokenizer::transition_from_GATHER_WORD(char first_char)
{
	token_string += first_char;
	while (position < input.length())
	{
		char c = input[position++];
		if (is_whitespace(c)) break;
		else token_string += c;
	}
	return Token(TokenType::WORD, token_string);
}